File: src/Spectra.cc
#include "pfs/drp/stella/Spectra.h"

namespace pfsDRPStella = pfs::drp::stella;

/** @brief Construct a Spectrum with empty vectors of specified size (default 0)
 */
template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
pfsDRPStella::Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>::Spectrum(size_t length,
                                                                           size_t iTrace) 
  : _length(length),
    _iTrace(iTrace),
    _isWavelengthSet(false)
{
  _spectrum = ndarray::allocate(length);
  _mask = ndarray::allocate(length);
  _variance = ndarray::allocate(length);
  _wavelength = ndarray::allocate(length);
}

template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
pfsDRPStella::Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>::Spectrum(Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT> const& spectrum,
                                                                           size_t iTrace) 
  : _length(spectrum.getLength()),
    _iTrace(spectrum.getITrace()),
    _isWavelengthSet(spectrum.isWavelengthSet())
{
  _spectrum = ndarray::allocate(spectrum.getSpectrum().getShape()[0]);
  _mask = ndarray::allocate(spectrum.getMask().getShape()[0]);
  _variance = ndarray::allocate(spectrum.getVariance().getShape()[0]);
  _wavelength = ndarray::allocate(spectrum.getWavelength().getShape()[0]);
  _spectrum.deep() = spectrum.getSpectrum();
  _mask.deep() = spectrum.getMask();
  _variance.deep() = spectrum.getVariance();
  _wavelength.deep() = spectrum.getWavelength();
  if (iTrace != 0)
    _iTrace = iTrace;
}

template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
bool pfsDRPStella::Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>::setSpectrum(const ndarray::Array<SpectrumT, 1, 1> & spectrum)
{
  /// Check length of input spectrum
  if (spectrum.getShape()[0] != _length){
    string message("pfsDRPStella::Spectrum::setSpectrum: ERROR: spectrum->size()=");
    message += to_string(spectrum.getShape()[0]) + string(" != _length=") + to_string(_length);
    cout << message << endl;
    throw LSST_EXCEPT(pexExcept::Exception, message.c_str());    
  }
  _spectrum.deep() = spectrum;
  return true;
}

template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
bool pfsDRPStella::Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>::setVariance(const ndarray::Array<VarianceT, 1, 1> & variance)
{
  /// Check length of input variance
  if (variance.getShape()[0] != _length){
    string message("pfsDRPStella::Spectrum::setVariance: ERROR: variance->size()=");
    message += to_string(variance.getShape()[0]) + string(" != _length=") + to_string(_length);
    cout << message << endl;
    throw LSST_EXCEPT(pexExcept::Exception, message.c_str());    
  }
  _variance.deep() = variance;
  return true;
}

template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
bool pfsDRPStella::Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>::setWavelength(const ndarray::Array<WavelengthT, 1, 1> & wavelength)
{
  /// Check length of input wavelength
  if (wavelength.getShape()[0] != _length){
    string message("pfsDRPStella::Spectrum::setWavelength: ERROR: wavelength->size()=");
    message += to_string(wavelength.getShape()[0]) + string(" != _length=") + to_string(_length);
    cout << message << endl;
    throw LSST_EXCEPT(pexExcept::Exception, message.c_str());    
  }
  _wavelength.deep() = wavelength;
  return true;
}

template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
bool pfsDRPStella::Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>::setMask(const ndarray::Array<MaskT, 1, 1> & mask)
{
  /// Check length of input mask
  if (mask.getShape()[0] != _length){
    string message("pfsDRPStella::Spectrum::setMask: ERROR: mask->size()=");
    message += to_string(mask.getShape()[0]) + string(" != _length=") + to_string(_length);
    cout << message << endl;
    throw LSST_EXCEPT(pexExcept::Exception, message.c_str());    
  }
  _mask.deep() = mask;
  return true;
}

template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
bool pfsDRPStella::Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>::setLength(const size_t length){
  pfsDRPStella::math::resize(_spectrum, length);
  pfsDRPStella::math::resize(_mask, length);
  pfsDRPStella::math::resize(_variance, length);
  pfsDRPStella::math::resize(_wavelength, length);
  if (length > _length){
    WavelengthT val = _wavelength[_length = 1];
    for (auto it = _wavelength.begin() + length; it != _wavelength.end(); ++it)
      *it = val;
  }
  _length = length;
  return true;
}

///SpectrumSet
template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
pfsDRPStella::SpectrumSet<SpectrumT, MaskT, VarianceT, WavelengthT>::SpectrumSet(size_t nSpectra, size_t length)
        : _spectra(new std::vector<PTR(pfsDRPStella::Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>)>())
{
  for (size_t i = 0; i < nSpectra; ++i){
    PTR(pfsDRPStella::Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>) spec(new pfsDRPStella::Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>(length));
    _spectra->push_back(spec);
    (*_spectra)[i]->setITrace(i);
  }
}
    
template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
pfsDRPStella::SpectrumSet<SpectrumT, MaskT, VarianceT, WavelengthT>::SpectrumSet(const PTR(std::vector<PTR(Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>)>) &spectrumVector)
        : _spectra(spectrumVector)
{}
//  for (int i = 0; i < spectrumVector->size(); ++i){
//    (*_spectra)[i]->setITrace(i);
//  }
//}
    
template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
bool pfsDRPStella::SpectrumSet<SpectrumT, MaskT, VarianceT, WavelengthT>::setSpectrum(const size_t i,     /// which spectrum?
                     const PTR(Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>) & spectrum /// the Spectrum for the ith aperture
                      )
{
  if (i > _spectra->size()){
    string message("SpectrumSet::setSpectrum(i=");
    message += to_string(i) + "): ERROR: i > _spectra->size()=" + to_string(_spectra->size());
    cout << message << endl;
    throw LSST_EXCEPT(pexExcept::Exception, message.c_str());
  }

  if (i == static_cast<int>(_spectra->size())){
    _spectra->push_back(spectrum);
  }
  else{
    (*_spectra)[i] = spectrum;
  }
  return true;
}

/// add one Spectrum to the set
template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
void pfsDRPStella::SpectrumSet<SpectrumT, MaskT, VarianceT, WavelengthT>::addSpectrum(Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT> const& spectrum /// the Spectrum to add
)
{
  PTR(Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>) ptr(new Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>(spectrum));
  _spectra->push_back(ptr);
  return;
}
template<typename SpectrumT, typename MaskT, typename VarianceT, typename WavelengthT>
void pfsDRPStella::SpectrumSet<SpectrumT, MaskT, VarianceT, WavelengthT>::addSpectrum(PTR(Spectrum<SpectrumT, MaskT, VarianceT, WavelengthT>) const& spectrum /// the Spectrum to add
)
{
  _spectra->push_back(spectrum);
  return;
}
  
template<typename ImageT, typename MaskT, typename VarianceT, typename WavelengthT>
PTR(pfsDRPStella::Spectrum<ImageT, MaskT, VarianceT, WavelengthT>)& pfsDRPStella::SpectrumSet<ImageT, MaskT, VarianceT, WavelengthT>::getSpectrum(const size_t i){
  if (i >= _spectra->size()){
    string message("SpectrumSet::getSpectrum(i=");
    message += to_string(i) + "): ERROR: i >= _spectra->size()=" + to_string(_spectra->size());
    cout << message << endl;
    throw LSST_EXCEPT(pexExcept::Exception, message.c_str());
  }
  return _spectra->at(i); 
}

template<typename ImageT, typename MaskT, typename VarianceT, typename WavelengthT>
PTR(const pfsDRPStella::Spectrum<ImageT, MaskT, VarianceT, WavelengthT>) const& pfsDRPStella::SpectrumSet<ImageT, MaskT, VarianceT, WavelengthT>::getSpectrum(const size_t i) const { 
  if (i >= _spectra->size()){
    string message("SpectrumSet::getSpectrum(i=");
    message += to_string(i) + "): ERROR: i >= _spectra->size()=" + to_string(_spectra->size());
    cout << message << endl;
    throw LSST_EXCEPT(pexExcept::Exception, message.c_str());
  }

  return _spectra->at(i); 
}

template<typename ImageT, typename MaskT, typename VarianceT, typename WavelengthT>
bool pfsDRPStella::SpectrumSet<ImageT, MaskT, VarianceT, WavelengthT>::erase(const size_t iStart, const size_t iEnd){
  if (iStart >= _spectra->size()){
    string message("SpectrumSet::erase(iStart=");
    message += to_string(iStart) + ", iEnd=" + to_string(iEnd) + "): ERROR: iStart >= _spectra->size()=" + to_string(_spectra->size());
    cout << message << endl;
    throw LSST_EXCEPT(pexExcept::Exception, message.c_str());
  }

  if (iEnd >= _spectra->size()){
    string message("SpectrumSet::erase(iStart=");
    message += to_string(iStart) + ", iEnd=" + to_string(iEnd) + "): ERROR: iEnd >= _spectra->size()=" + to_string(_spectra->size());
    cout << message << endl;
    throw LSST_EXCEPT(pexExcept::Exception, message.c_str());
  }

  if (iEnd > 0){
    if (iStart > iEnd){
      string message("SpectrumSet::erase(iStart=");
      message += to_string(iStart) + ", iEnd=" + to_string(iEnd) + "): ERROR: iStart > iEnd";
      cout << message << endl;
      throw LSST_EXCEPT(pexExcept::Exception, message.c_str());
    }
  }
  if (iStart == (_spectra->size()-1)){
    _spectra->pop_back();
  }
  else{
    if (iEnd == 0)
      _spectra->erase(_spectra->begin() + iStart);
    else
      _spectra->erase(_spectra->begin() + iStart, _spectra->begin() + iEnd);
  }
  return true;
}


template<typename T>
PTR(T) pfsDRPStella::utils::getPointer(T &obj){
  PTR(T) pointer(new T(obj));
  return pointer;
}

//template class pfsDRPStella::Spectrum<float>;
//template class pfsDRPStella::Spectrum<double>;
template class pfsDRPStella::Spectrum<float, unsigned int, float, float>;
template class pfsDRPStella::Spectrum<double, unsigned int, float, float>;
template class pfsDRPStella::Spectrum<float, unsigned short, float, float>;
template class pfsDRPStella::Spectrum<double, unsigned short, float, float>;
//template class pfsDRPStella::Spectrum<float, unsigned int, double, double>;
//template class pfsDRPStella::Spectrum<double, unsigned int, double, double>;
//template class pfsDRPStella::Spectrum<float, unsigned short, double, double>;
//template class pfsDRPStella::Spectrum<double, unsigned short, double, double>;

//template class pfsDRPStella::SpectrumSet<float>;
//template class pfsDRPStella::SpectrumSet<double>;
template class pfsDRPStella::SpectrumSet<float, unsigned int, float, float>;
template class pfsDRPStella::SpectrumSet<double, unsigned int, float, float>;
template class pfsDRPStella::SpectrumSet<float, unsigned short, float, float>;
template class pfsDRPStella::SpectrumSet<double, unsigned short, float, float>;
//template class pfsDRPStella::SpectrumSet<float, unsigned int, double, double>;
//template class pfsDRPStella::SpectrumSet<double, unsigned int, double, double>;
//template class pfsDRPStella::SpectrumSet<float, unsigned short, double, double>;
//template class pfsDRPStella::SpectrumSet<double, unsigned short, double, double>;

template PTR(afwImage::MaskedImage<float, unsigned short, float>) pfsDRPStella::utils::getPointer(afwImage::MaskedImage<float, unsigned short, float> &);
template PTR(afwImage::MaskedImage<double, unsigned short, float>) pfsDRPStella::utils::getPointer(afwImage::MaskedImage<double, unsigned short, float> &);
template PTR(std::vector<unsigned short>) pfsDRPStella::utils::getPointer(std::vector<unsigned short> &);
template PTR(std::vector<unsigned int>) pfsDRPStella::utils::getPointer(std::vector<unsigned int> &);
template PTR(std::vector<int>) pfsDRPStella::utils::getPointer(std::vector<int> &);
template PTR(std::vector<float>) pfsDRPStella::utils::getPointer(std::vector<float> &);
template PTR(std::vector<double>) pfsDRPStella::utils::getPointer(std::vector<double> &);
template PTR(pfsDRPStella::Spectrum<float, unsigned short, float, float>) pfsDRPStella::utils::getPointer(pfsDRPStella::Spectrum<float, unsigned short, float, float> &);
template PTR(pfsDRPStella::Spectrum<double, unsigned short, float, float>) pfsDRPStella::utils::getPointer(pfsDRPStella::Spectrum<double, unsigned short, float, float> &);

File: tests/Spectra.py
#!/usr/bin/env python
"""
Tests for measuring things

Run with:
   python Spectra.py
or
   python
   >>> import Spectra; Spectra.run()
"""

import unittest
import sys
import numpy as np
import lsst.utils.tests as tests
import lsst.afw.image as afwImage
import lsst.afw.geom as afwGeom
import pfs.drp.stella as drpStella
import lsst.afw.display.ds9 as ds9
import lsst.afw.display.utils as displayUtils

try:
    type(display)
except NameError:
    display = False

class SpectraTestCase(tests.TestCase):
    """A test case for measuring Spectra quantities"""

    def setUp(self):
        latest = True
        if latest:
            flatfile = "sampledFlatx2-IR-0-23.fits"
            combfile = "sampledCombx2-IR-0-23.fits"
        else:
            flatfile = "/Users/azuri/spectra/pfs/2014-10-28/sampledFlatx2-IR-0-23-5-10-nonoise.fits"
            combfile = "/Users/azuri/spectra/pfs/2014-10-28/sampledCombx2-IR-0-23-5-10-nonoise.fits"
        self.flat = afwImage.ImageF(flatfile)
        self.flat = afwImage.makeExposure(afwImage.makeMaskedImage(self.flat))
        self.bias = afwImage.ImageF(flatfile, 3)
        self.flat.getMaskedImage()[:] -= self.bias

        self.comb = afwImage.ImageF(combfile)
        self.comb = afwImage.makeExposure(afwImage.makeMaskedImage(self.comb))
        bias = afwImage.ImageF(combfile, 3)
        self.comb.getMaskedImage()[:] -= bias
        
        self.ftffc = drpStella.FiberTraceFunctionFindingControl()
        self.ftffc.fiberTraceFunctionControl.order = 4
        
        self.ftpfc = drpStella.FiberTraceProfileFittingControl()

        del bias
        del flatfile
        del combfile
        del latest
        
    def tearDown(self):
        del self.flat
        del self.comb
        del self.bias
        del self.ftffc
        del self.ftpfc

    def testSpectrumConstructors(self):
        """Test that we can create a Spectrum with the standard constructor"""
        spec = drpStella.SpectrumF()
        self.assertEqual(spec.getLength(), 0)
        self.assertEqual(spec.getITrace(), 0)
        
        length = 10
        iTrace = 2
        spec = drpStella.SpectrumF(length, iTrace)
        self.assertEqual(spec.getLength(), length)
        self.assertEqual(spec.getITrace(), iTrace)
        
        """Test copy constructor"""
        specCopy = drpStella.SpectrumF(spec)
        self.assertEqual(specCopy.getLength(), length)
        self.assertEqual(specCopy.getITrace(), iTrace)
        
    def testSpectrumMethods(self):
        """Test getSpectrum"""
        size = 100
        spec = drpStella.SpectrumF(size)
        vec = spec.getSpectrum()
        self.assertEqual(vec.shape[0], size)
        
        """Test setSpectrum"""
        """Test that we can assign a spectrum of the correct length"""
        vecf = drpStella.indGenNdArrF(size)
        self.assertTrue(spec.setSpectrum(vecf))
        self.assertEqual(spec.getSpectrum()[3], vecf[3])
        
        """Test that we can't assign a spectrum of the wrong length"""
        vecf = drpStella.indGenNdArrF(size+1)
        try:
            spec.setSpectrum(vecf)
        except:
            e = sys.exc_info()[1]
            message = str.split(e.message, "\n")
            for i in range(len(message)):
                print "element",i,": <",message[i],">"
            expected = "pfsDRPStella::Spectrum::setSpectrum: ERROR: spectrum->size()="+str(vecf.shape[0])+" != _length="+str(spec.getLength())
            self.assertEqual(message[0],expected)
        self.assertEqual(spec.getSpectrum().shape[0], size)
        
        """Test getVariance"""
        vec = spec.getVariance()
        self.assertEqual(vec.shape[0], size)
        
        """Test setVariance"""
        """Test that we can assign a variance vector of the correct length"""
        vecf = drpStella.indGenNdArrF(size)
        self.assertTrue(spec.setVariance(vecf))
        self.assertEqual(spec.getVariance()[3], vecf[3])
        
        """Test that we can't assign a variance vector of the wrong length"""
        vecf = drpStella.indGenNdArrF(size+1)
        try:
            self.assertFalse(spec.setVariance(vecf))
        except:
            e = sys.exc_info()[1]
            message = str.split(e.message, "\n")
            for i in range(len(message)):
                print "element",i,": <",message[i],">"
            expected = "pfsDRPStella::Spectrum::setVariance: ERROR: variance->size()="+str(vecf.shape[0])+" != _length="+str(spec.getLength())
            self.assertEqual(message[0],expected)
        self.assertEqual(spec.getVariance().shape[0], size)
        
        """Test getWavelength"""
        vec = spec.getWavelength()
        self.assertEqual(vec.shape[0], size)
        
        """Test setWavelength"""
        """Test that we can assign a wavelength vector of the correct length"""
        vecf = drpStella.indGenNdArrF(size)
        self.assertTrue(spec.setWavelength(vecf))
        self.assertEqual(spec.getWavelength()[3], vecf[3])
        
        """Test that we can't assign a wavelength vector of the wrong length"""
        vecf = drpStella.indGenNdArrF(size+1)
        try:
            self.assertFalse(spec.setWavelength(vecf))
        except:
            e = sys.exc_info()[1]
            message = str.split(e.message, "\n")
            for i in range(len(message)):
                print "element",i,": <",message[i],">"
            expected = "pfsDRPStella::Spectrum::setWavelength: ERROR: wavelength->size()="+str(vecf.shape[0])+" != _length="+str(spec.getLength())
            self.assertEqual(message[0],expected)
        self.assertEqual(spec.getWavelength().shape[0], size)
        
        
        """Test getMask"""
        vec = spec.getMask()
        print "vec = ",vec
        print "dir(vec): ",dir(vec)
        self.assertEqual(vec.shape[0], size)
        
        """Test setMask"""
        """Test that we can assign a mask vector of the correct length"""
        vecf = drpStella.indGenNdArrUS(size)
        self.assertTrue(spec.setMask(vecf))
        self.assertEqual(spec.getMask()[3], vecf[3])
        
        """Test that we can't assign a mask vector of the wrong length"""
        vecus = drpStella.indGenNdArrUS(size+1)
        try:
            self.assertFalse(spec.setMask(vecus))
        except:
            e = sys.exc_info()[1]
            message = str.split(e.message, "\n")
            for i in range(len(message)):
                print "element",i,": <",message[i],">"
            expected = "pfsDRPStella::Spectrum::setMask: ERROR: mask->size()="+str(vecus.shape[0])+" != _length="+str(spec.getLength())
            self.assertEqual(message[0],expected)
        self.assertEqual(spec.getMask().shape[0], size)

        if False: 
            """Test setLength"""
            """If newLength < oldLength, vectors are supposed to be cut off, otherwise ZEROs are appended to the end of the vectors (last wavelength value for wavelength vector)"""
            """Test same size"""
            vecf = drpStella.indGenNdArrF(size)
            vecus = drpStella.indGenNdArrUS(size)
            self.assertTrue(spec.setLength(size))
            self.assertEqual(spec.getLength(), size)
            self.assertEqual(spec.getSpectrum()[size-1], vecf[size-1])
            self.assertEqual(spec.getSpectrum().shape[0], size)
            self.assertEqual(spec.getVariance().shape[0], size)
            self.assertEqual(spec.getVariance()[size-1], vecf[size-1])
            self.assertEqual(spec.getMask().shape[0], size)
            self.assertEqual(spec.getMask()[size-1], vecus[size-1])
            self.assertEqual(spec.getWavelength().shape[0], size)
            self.assertEqual(spec.getWavelength()[size-1], vecf[size-1])

            """Test longer size"""
            self.assertTrue(spec.setLength(size+1))
            self.assertEqual(spec.getLength(), size+1)
            self.assertEqual(spec.getSpectrum()[size], 0)
            self.assertEqual(spec.getSpectrum().shape[0], size+1)
            self.assertEqual(spec.getVariance().shape[0], size+1)
            self.assertEqual(spec.getVariance()[size], 0)
            self.assertEqual(spec.getMask().shape[0], size+1)
            self.assertEqual(spec.getMask()[size], 0)
            self.assertEqual(spec.getWavelength().shape[0], size+1)
            self.assertEqual(spec.getWavelength()[size], vecf[size-1])

            """Test shorter size"""
            self.assertTrue(spec.setLength(size-1))
            self.assertEqual(spec.getLength(), size-1)
            self.assertEqual(spec.getSpectrum()[size-2], vecf[size-2])
            self.assertEqual(spec.getSpectrum().shape[0], size-1)
            self.assertEqual(spec.getVariance().shape[0], size-1)
            self.assertEqual(spec.getVariance()[size-2], vecf[size-2])
            self.assertEqual(spec.getMask().shape[0], size-1)
            self.assertEqual(spec.getMask()[size-2], vecus[size-2])
            self.assertEqual(spec.getWavelength().shape[0], size-1)
            self.assertEqual(spec.getWavelength()[size-2], vecf[size-2])
        
        """Test get/setITrace"""
        self.assertEqual(spec.getITrace(), 0)
        spec.setITrace(10)
        self.assertEqual(spec.getITrace(), 10)
        
        """Test isWaveLengthSet"""
        self.assertFalse(spec.isWavelengthSet())
            
    def testSpectrumSetConstructors(self):
        """Test SpectrumSetConstructors"""
        """Test Standard Constructor"""
        specSet = drpStella.SpectrumSetF()
        self.assertEqual(specSet.size(), 0)
        
        size = 3
        specSet = drpStella.SpectrumSetF(size)
        self.assertEqual(specSet.size(), size)
        for i in range(size):
            self.assertEqual(specSet.getSpectrum(i).getSpectrum().shape[0], 0)
        
        length = 33
        specSet = drpStella.SpectrumSetF(size, length)
        self.assertEqual(specSet.size(), size)
        for i in range(size):
            self.assertEqual(specSet.getSpectrum(i).getSpectrum().shape[0], length)
            
        """Test copy constructor"""
        specSetCopy = drpStella.SpectrumSetF(specSet)
        self.assertEqual(specSetCopy.size(), specSet.size())
        for i in range(specSet.size()):
            self.assertEqual(specSetCopy.getSpectrum(i).getLength(), specSet.getSpectrum(i).getLength())
        if False:
            val = 3.3
            pos = 3
            vecf = list(drpStella.indGenF(length))
            print type(vecf)
            print vecf
            print dir(vecf)
            vecf[pos] = val
            pvecf = drpStella.SpecVectorF(vecf)
            self.assertTrue(specSet.getSpectrum(0).setSpectrum(pvecf))
            specSetCopy.getSpectrum(0).setSpectrum(pvecf)
            self.assertAlmostEqual(specSetCopy.getSpectrum(i).getSpectrum()[pos], val)
            
        """Test constructor from vector of spectra"""
        specSetV = drpStella.SpectrumSetF(specSet.getSpectra())
        self.assertEqual(specSet.size(), specSetV.size())
        for i in range(specSet.size()):
            self.assertEqual(specSetV.getSpectrum(i).getITrace(), i)
            
    def testSpectrumSetAddSetErase(self):
        size = 3
        length = 100
        specSet = drpStella.SpectrumSetF(size, length)
        spec = drpStella.SpectrumF(length)
        specNew = drpStella.SpectrumF(length+1)
        
        """Test that we cannot set a spectrum outside the limits 0 <= pos <= size"""
        try:
            specSet.setSpectrum(-1, specNew)
        except:
            e = sys.exc_info()[1]
            message = str.split(e.message, "\n")
            for i in range(len(message)):
                print "element",i,": <",message[i],">"
            expected = "in method 'SpectrumSetF_setSpectrum', argument 2 of type 'size_t'"
            self.assertEqual(message[0],expected)
        try:
            specSet.setSpectrum(size+1, specNew)
        except:
            e = sys.exc_info()[1]
            message = str.split(e.message, "\n")
            for i in range(len(message)):
                print "element",i,": <",message[i],">"
            expected = "SpectrumSet::setSpectrum(i="+str(size+1)+"): ERROR: i > _spectra->size()="+str(size)
            self.assertEqual(message[0],expected)
            
        """Test that we can set/add a spectrum"""
        self.assertTrue(specSet.setSpectrum(size-1, specNew))
        self.assertEqual(specSet.size(), size)
        self.assertTrue(specSet.setSpectrum(size, specNew))
        self.assertEqual(specSet.size(), size+1)
        
        specSet.addSpectrum(specNew)
        self.assertEqual(specSet.size(), size+2)
        
        """Test that we can't erase spectra outside the limits"""
        size = specSet.size()
        try:
            specSet.erase(size)
        except:
            e = sys.exc_info()[1]
            message = str.split(e.message, "\n")
            for i in range(len(message)):
                print "element",i,": <",message[i],">"
            expected = "SpectrumSet::erase(iStart="+str(size)+", iEnd=0): ERROR: iStart >= _spectra->size()="+str(size)
            self.assertEqual(message[0],expected)

        try:
            specSet.erase(size-1, size)
        except:
            e = sys.exc_info()[1]
            message = str.split(e.message, "\n")
            for i in range(len(message)):
                print "element",i,": <",message[i],">"
            expected = "SpectrumSet::erase(iStart="+str(size-1)+", iEnd="+str(size)+"): ERROR: iEnd >= _spectra->size()="+str(size)
            self.assertEqual(message[0],expected)

        try:
            specSet.erase(2, 1)
        except:
            e = sys.exc_info()[1]
            message = str.split(e.message, "\n")
            for i in range(len(message)):
                print "element",i,": <",message[i],">"
            expected = "SpectrumSet::erase(iStart=2, iEnd=1): ERROR: iStart > iEnd"
            self.assertEqual(message[0],expected)

        try:
            specSet.erase(-1)
        except:
            e = sys.exc_info()[1]
            message = str.split(e.message, "\n")
            for i in range(len(message)):
                print "element",i,": <",message[i],">"
            expected = "Wrong number or type of arguments for overloaded function 'SpectrumSetF_erase'."
            self.assertEqual(message[0],expected)

        """Test that we CAN erase spectra inside the limits"""
        self.assertTrue(specSet.erase(size-1))
        self.assertEqual(specSet.size(), size-1)
        
        self.assertTrue(specSet.erase(0, 1))
        self.assertEqual(specSet.size(), size-2)

        self.assertTrue(specSet.erase(0,2))
        self.assertEqual(specSet.size(), size-4)
        
    def testGetSpectra(self):
        """test getSpectra"""
        size = 3
        length = 100
        specSet = drpStella.SpectrumSetF(size,length)
        self.assertEqual(specSet.getSpectra()[0].getSpectrum().shape[0], length)
            
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

def suite():
    """Returns a suite containing all the test cases in this module."""
    tests.init()

    suites = []
    suites += unittest.makeSuite(SpectraTestCase)
#    suites += unittest.makeSuite(tests.MemoryTestCase)
    return unittest.TestSuite(suites)

def run(exit = False):
    """Run the tests"""
    tests.run(suite(), exit)

if __name__ == "__main__":
    from argparse import ArgumentParser
    parser = ArgumentParser()
    parser.add_argument("--display", '-d', default=False, action="store_true", help="Activate display?")
    parser.add_argument("--verbose", '-v', type=int, default=0, help="Verbosity level")
    args = parser.parse_args()
    display = args.display
    verbose = args.verbose
    run(True)
